### PR TITLE
Fixing unicode urls in python3

### LINKED ISSUE #4033: Books with non-ascii titles erroring when clicked from search
### Steps to Reproduce

1. Go to https://openlibrary.org/search?q=h%C3%A9las&mode=everything
2. Click on any book with a non-ascii character in the title

* Actual: 500 internal error on e.g. https://openlibrary.org/works/OL11565520W?edition=
* Expected: See the book

### Details

- **Logged in (Y/N)?** Y
- **Browser type/version?** FF82
- **Operating system?** Win10
- **Environment (prod/dev/local)?** prod

### Stakeholders
@cclauss

### PR DESCRIPTION
**Pull request by mekarpeles**
Closes #4033 

- avoids clobbering `property` and `type` keywords

### Technical

Uses urlencoding to get around unicode characters being encoded as invalid latin characters by webpy application.py

### Testing

Tested on staging (works on python3) -- breaks on dev (python2)

### CODE REVIEW

**Comment by mekarpeles** [comment]
addresses https://github.com/webpy/webpy/issues/687

File: openlibrary/core/processors/readableurls.py [hunk 1]
```diff
@@ -161,9 +165,13 @@ def match(path):
     if thing:
         prefix = thing.key
 
-    if thing and thing.type.key == type:
-        title = thing.get(property) or default_title
-        middle = '/' + h.urlsafe(title.strip())
+    if thing and thing.type.key == _type:
+        title = thing.get(_property) or default_title
+        try:
+            from urllib.parse import quote_plus
```
Comment (cclauss) [inline openlibrary/core/processors/readableurls.py:171]:
```suggestion
            from six.moves.urllib_parse import quote_plus
```
And please move the import to the top of the file.

**Review by cclauss — APPROVED** [review]
LGTM

**Comment by LeadSongDog** [comment]
Please be sure that fixing this does not reawaken the issues  #149 #178 #315 #714 #2102 #4033. Frankly, I still see no value in having text in the slug. It has been buggy for years, costs performance, and exposes titles in browser history. It would be far cleaner to use only the numeric identifiers for authors, works, and editions in the respective URLs.

**Comment by mekarpeles** [comment]
@LeadSongDog according to experts I've spoken with on the topic, there is significant SEO value in including a human readable "slug" in the URL. In addition, there is further SEO cost in switching all of these urls to a redirect. This PR closes a regression. If we want to do something different, we can open an issue and discuss the merits of such a change. As an archive, my first priority is making sure our urls work as they are currently expected to.

I don't see why this should affect the other issues you listed. If these are issues, they are likely to emerge as issues for no better reason that we are moving from python2->3 and strings/bytes/unicode work differently. And we're well aware that regressions will occur as we inevitably encounter and address these differences.

File: openlibrary/core/processors/readableurls.py [hunk 2]
```diff
@@ -161,9 +165,13 @@ def match(path):
     if thing:
         prefix = thing.key
 
-    if thing and thing.type.key == type:
-        title = thing.get(property) or default_title
-        middle = '/' + h.urlsafe(title.strip())
+    if thing and thing.type.key == _type:
+        title = thing.get(_property) or default_title
+        try:
+            from six.moves.urllib_parse import quote_plus
```
Comment (mekarpeles) [inline openlibrary/core/processors/readableurls.py:171]: On second thought -- this needs to fail on python2 (the solution should be conditional on python3, thus this is basically a py3 check)

```suggestion
           from urllib.parse import quote_plus
```

Comment (cclauss) [reply openlibrary/core/processors/readableurls.py:171]: OK.  Now it makes sense.  Perhaps add a comment to explain.

File: openlibrary/core/processors/readableurls.py [hunk 3]
```diff
@@ -161,9 +165,14 @@ def match(path):
     if thing:
         prefix = thing.key
 
-    if thing and thing.type.key == type:
-        title = thing.get(property) or default_title
-        middle = '/' + h.urlsafe(title.strip())
+    if thing and thing.type.key == _type:
+        title = thing.get(_property) or default_title
+        try:
+            # Explicitly only run for python3 to solve #4033
```
Comment (cclauss) [inline openlibrary/core/processors/readableurls.py:171]:
```suggestion
            # Explicitly only run for python3 to solve #4033
```

**Pull request merged**